File: LeetCode_Algorithm/735(Asteroid Collision)/735.js
var asteroidCollision = function(asteroids) {
  // sol1 O(n) O(n) 60ms 99%
  let ans = []
  for (let asteroid of asteroids) {
    let flag = true
    while (ans.length > 0 && asteroid < 0 && 0 < ans[ans.length-1]) {
      if (ans[ans.length-1] < -asteroid) {
        ans.pop(); 
        continue
      } else if (ans[ans.length-1] == -asteroid) {
        ans.pop(); flag = false
      } else {
        flag = false
      }
      break
    }
    if (flag == true) {
      ans.push(asteroid)
    }
  }
  return ans
};

console.log(asteroidCollision([5,10,-5]))

File: LeetCode_Algorithm/735(Asteroid Collision)/735.py
from typing import List

class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        # sol1 O(n) O(n) 104ms 87% 17.5MB 84$%
        ans = []
        for asteroid in asteroids:
            while ans and asteroid < 0 < ans[-1]:
                if ans[-1] < -asteroid:
                    ans.pop(); 
                    continue
                elif ans[-1] == -asteroid:
                    ans.pop(); 
                break
            else:
                ans.append(asteroid)
        return ans
    


print(Solution().asteroidCollision([5,10,-5]))
print(Solution().asteroidCollision([5,10,-11]))
print(Solution().asteroidCollision([1,2,3]))
print(Solution().asteroidCollision([-4,2,3]))
print(Solution().asteroidCollision([-4,4]))



